hbase-server/src/main/java/org/apache/hadoop/hbase/monitoring/ThreadLocalServerSideScanMetrics.java

```diff
@@ -0,0 +1,180 @@
+ * server design: on the server side, the thread that opens a {@link RegionScanner} and calls
+ * {@link RegionScanner#nextRaw(java.util.List, ScannerContext)} is the same thread that reads HFile
+ * blocks. This design allows thread-local storage to effectively capture metrics without
+ * cross-thread synchronization.
+ * <h3>Special Handling for Parallel Operations</h3> The only deviation from the single-thread model
+ * occurs when {@link org.apache.hadoop.hbase.regionserver.handler.ParallelSeekHandler} is used for
+ * parallel store file seeking. In this case, special handling ensures that metrics are captured
+ * correctly across multiple threads. The
+ * {@link org.apache.hadoop.hbase.regionserver.handler.ParallelSeekHandler} captures metrics from
+ * worker threads and aggregates them back to the main scan thread. Please refer to the javadoc of
+ * {@link org.apache.hadoop.hbase.regionserver.handler.ParallelSeekHandler} for detailed information
+ * about this parallel processing mechanism.
+ * <h3>Usage Pattern</h3>
+ * <ol>
+ * <li>Enable metrics collection: {@link #setScanMetricsEnabled(boolean)}</li>
+ * <li>Reset counters at scan start: {@link #reset()}</li>
+ * <li>Increment counters during I/O operations using the various {@code add*} methods</li>
+ * <li>Populate the main metrics object:
+ * {@link #populateServerSideScanMetrics(ServerSideScanMetrics)}</li>
+ * </ol>
+ * <h3>Thread Safety</h3> This class is thread-safe. Each thread maintains its own set of counters
+ * through {@link ThreadLocal} storage, ensuring that metrics from different scan operations do not
+ * interfere with each other.
+ * @see ServerSideScanMetrics
+ * @see RegionScanner
+ * @see org.apache.hadoop.hbase.regionserver.handler.ParallelSeekHandler
+ */
+@InterfaceAudience.Private
+public final class ThreadLocalServerSideScanMetrics {
+  private ThreadLocalServerSideScanMetrics() {
+  }
+
+  private static final ThreadLocal<Boolean> isScanMetricsEnabled = new ThreadLocal<>() {
+    @Override
+    protected Boolean initialValue() {
+      return false;
+    }
+  };
+
+  private static final ThreadLocal<AtomicLong> bytesReadFromFs = new ThreadLocal<>() {
+    @Override
+    protected AtomicLong initialValue() {
+      return new AtomicLong(0);
+    }
+  };
+
+  private static final ThreadLocal<AtomicLong> bytesReadFromBlockCache = new ThreadLocal<>() {
+    @Override
+    protected AtomicLong initialValue() {
+      return new AtomicLong(0);
+    }
+  };
+
+  private static final ThreadLocal<AtomicLong> bytesReadFromMemstore = new ThreadLocal<>() {
+    @Override
+    protected AtomicLong initialValue() {
+      return new AtomicLong(0);
+    }
+  };
+
+  private static final ThreadLocal<AtomicLong> blockReadOpsCount = new ThreadLocal<>() {
+    @Override
+    protected AtomicLong initialValue() {
+      return new AtomicLong(0);
+    }
+  };
+
+  public static final void setScanMetricsEnabled(boolean enable) {
```

Comment:
`final` is not required for any of these methods, let's remove it. For above constants, it's fine to have it.